File: src/zigrix/evidence.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from zigrix.events import append_event, now_iso
from zigrix.paths import ZigrixPaths, ensure_project_state
from zigrix.state import load_task, rebuild_index
from zigrix.worker import resolve_required_agents



def collect_evidence(
    paths: ZigrixPaths,
    *,
    task_id: str,
    agent_id: str,
    run_id: str = "",
    unit_id: str | None = None,
    session_key: str = "",
    session_id: str = "",
    transcript: str = "",
    summary: str = "",
    tool_results: list[str] | None = None,
    notes: str = "",
    limit: int = 40,
) -> dict[str, Any] | None:
    ensure_project_state(paths)
    task = load_task(paths, task_id)
    if not task:
        return None

    transcript_rows: list[dict[str, Any]] = []
    transcript_path = ""
    extracted: dict[str, Any] = {}
    if transcript:
        transcript_path = str(Path(transcript).expanduser().resolve())
        transcript_rows = _read_transcript(Path(transcript_path), limit=limit)
        extracted = _extract_evidence(transcript_rows)
    if summary:
        extracted["summary"] = summary
        extracted.setdefault("lastAssistant", summary)
    if tool_results:
        extracted["toolResults"] = list(tool_results)
    if notes:
        extracted["notes"] = notes

    out_dir = paths.evidence_dir / task_id
    out_dir.mkdir(parents=True, exist_ok=True)
    out_path = out_dir / f"{agent_id}.json"
    payload = {
        "ts": now_iso(),
        "taskId": task_id,
        "agentId": agent_id,
        "unitId": unit_id,
        "runId": run_id,
        "sessionKey": session_key or None,
        "sessionId": session_id or None,
        "transcriptPath": transcript_path or None,
        "evidence": extracted,
    }
    out_path.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    append_event(
        paths.events_file,
        {
            "event": "evidence_collected",
            "taskId": task_id,
            "phase": "verification",
            "actor": agent_id,
            "status": "IN_PROGRESS",
            "unitId": unit_id,
            "sessionKey": session_key or None,
            "payload": {
                "agentId": agent_id,
                "runId": run_id,
                "evidencePath": str(out_path),
            },
        },
    )
    rebuild_index(paths)
    return {
        "ok": True,
        "taskId": task_id,
        "agentId": agent_id,
        "evidencePath": str(out_path),
        "sessionId": session_id or None,
        "unitId": unit_id,
    }



def merge_evidence(
    paths: ZigrixPaths,
    *,
    task_id: str,
    required_agents: list[str] | None = None,
    require_qa: bool = False,
) -> dict[str, Any] | None:
    task = load_task(paths, task_id)
    if not task:
        return None

    task_dir = paths.evidence_dir / task_id
    task_dir.mkdir(parents=True, exist_ok=True)
    files = sorted(f for f in task_dir.glob("*.json") if f.name != "_merged.json")

    merged_items: list[dict[str, Any]] = []
    present: list[str] = []
    for file in files:
        try:
            data = json.loads(file.read_text(encoding="utf-8"))
        except Exception:
            continue
        if not isinstance(data, dict):
            continue
        agent_id = str(data.get("agentId") or file.stem)
        present.append(agent_id)
        merged_items.append(
            {
                "agentId": agent_id,
                "unitId": data.get("unitId"),
                "runId": data.get("runId"),
                "sessionKey": data.get("sessionKey"),
                "sessionId": data.get("sessionId"),
                "transcriptPath": data.get("transcriptPath"),
                "evidence": data.get("evidence", {}),
            }
        )

    required = list(required_agents or resolve_required_agents(task))
    missing = [agent for agent in required if agent not in present]
    qa_present = "qa-zig" in present
    complete = len(missing) == 0 and (not require_qa or qa_present)

    merged = {
        "ts": now_iso(),
        "taskId": task_id,
        "requiredAgents": required,
        "presentAgents": sorted(set(present)),
        "missingAgents": missing,
        "qaPresent": qa_present,
        "complete": complete,
        "items": merged_items,
    }
    out_path = task_dir / "_merged.json"
    out_path.write_text(json.dumps(merged, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    append_event(
        paths.events_file,
        {
            "event": "evidence_merged",
            "taskId": task_id,
            "phase": "verification",
            "actor": "zigrix",
            "status": "DONE_PENDING_REPORT" if complete else "IN_PROGRESS",
            "payload": {
                "requiredAgents": required,
                "missingAgents": missing,
                "complete": complete,
                "mergedPath": str(out_path),
                "qaPresent": qa_present,
            },
        },
    )
    rebuild_index(paths)
    return {
        "ok": True,
        "taskId": task_id,
        "complete": complete,
        "missingAgents": missing,
        "mergedPath": str(out_path),
    }



def _read_transcript(path: Path, *, limit: int = 40) -> list[dict[str, Any]]:
    if not path.exists():
        return []
    rows: list[dict[str, Any]] = []
    for line in path.read_text(encoding="utf-8", errors="ignore").splitlines()[-limit:]:
        if not line.strip():
            continue
        try:
            parsed = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(parsed, dict):
            rows.append(parsed)
    return rows



def _extract_evidence(rows: list[dict[str, Any]]) -> dict[str, Any]:
    last_assistant = None
    tool_results: list[Any] = []
    for row in rows:
        role = row.get("role")
        content = row.get("content")
        if role == "assistant" and content:
            last_assistant = content
        if role == "toolResult":
            tool_results.append(content)
    return {
        "lastAssistant": last_assistant,
        "toolResults": tool_results[-3:],
    }

File: src/orchestration/evidence.ts
import fs from 'node:fs';
import path from 'node:path';

import { appendEvent, nowIso } from '../state/events.js';
import { type ZigrixPaths, ensureBaseState } from '../state/paths.js';
import { type ZigrixTask, loadTask, rebuildIndex } from '../state/tasks.js';
import { resolveRequiredAgents } from './worker.js';

export type VerificationMapping = {
  dod: string;
  test: string;
};

function readTranscript(transcriptPath: string, limit = 40): Array<Record<string, unknown>> {
  if (!transcriptPath || !fs.existsSync(transcriptPath)) return [];
  return fs.readFileSync(transcriptPath, 'utf8')
    .split(/\r?\n/)
    .filter(Boolean)
    .slice(-limit)
    .flatMap((line) => {
      try {
        const parsed = JSON.parse(line) as Record<string, unknown>;
        return [parsed];
      } catch {
        return [];
      }
    });
}

function extractEvidence(rows: Array<Record<string, unknown>>): Record<string, unknown> {
  let lastAssistant: unknown = null;
  const toolResults: unknown[] = [];
  for (const row of rows) {
    if (row.role === 'assistant' && row.content) lastAssistant = row.content;
    if (row.role === 'toolResult') toolResults.push(row.content);
  }
  return { lastAssistant, toolResults: toolResults.slice(-3) };
}

function resolveQaAgentId(task: ZigrixTask): string | null {
  if (typeof task.qaAgentId === 'string' && task.qaAgentId.trim().length > 0) {
    return task.qaAgentId.trim();
  }

  const roleMap = task.roleAgentMap;
  if (roleMap && typeof roleMap === 'object') {
    const qaAgents = (roleMap as Record<string, unknown>).qa;
    if (Array.isArray(qaAgents) && qaAgents.length > 0) {
      const fromRoleMap = String(qaAgents[0]).trim();
      if (fromRoleMap.length > 0) return fromRoleMap;
    }
  }

  const required = Array.isArray(task.requiredAgents) ? task.requiredAgents : [];
  if (required.length === 1) {
    const only = String(required[0]).trim();
    if (only.length > 0) return only;
  }

  const orchestratorId = typeof task.orchestratorId === 'string' && task.orchestratorId.trim().length > 0
    ? task.orchestratorId.trim()
    : null;
  const fallback = required.find((agentId) => String(agentId).trim().length > 0 && String(agentId).trim() !== orchestratorId);
  return fallback ? String(fallback).trim() : null;
}

function sanitizeVerificationMapping(mapping: VerificationMapping): VerificationMapping | null {
  const dod = mapping.dod.trim();
  const test = mapping.test.trim();
  if (!dod || !test) return null;
  return { dod, test };
}

function collectVerificationMappings(evidence: Record<string, unknown>): VerificationMapping[] {
  const verification = evidence.verification;
  if (!verification || typeof verification !== 'object' || Array.isArray(verification)) return [];
  const rawMappings = (verification as Record<string, unknown>).mappings;
  if (!Array.isArray(rawMappings)) return [];
  return rawMappings.flatMap((item) => {
    if (!item || typeof item !== 'object' || Array.isArray(item)) return [];
    const dod = typeof (item as Record<string, unknown>).dod === 'string' ? String((item as Record<string, unknown>).dod) : '';
    const test = typeof (item as Record<string, unknown>).test === 'string' ? String((item as Record<string, unknown>).test) : '';
    const sanitized = sanitizeVerificationMapping({ dod, test });
    return sanitized ? [sanitized] : [];
  });
}

export function collectEvidence(paths: ZigrixPaths, params: {
  taskId: string;
  agentId: string;
  runId?: string;
  unitId?: string;
  sessionKey?: string;
  sessionId?: string;
  transcript?: string;
  summary?: string;
  toolResults?: string[];
  notes?: string;
  dodItems?: string[];
  testCases?: string[];
  verificationMappings?: VerificationMapping[];
  limit?: number;
}): Record<string, unknown> | null {
  ensureBaseState(paths);
  const task = loadTask(paths, params.taskId);
  if (!task) return null;
  const transcriptRows = params.transcript ? readTranscript(path.resolve(params.transcript), params.limit ?? 40) : [];
  const extracted = extractEvidence(transcriptRows);
  if (params.summary) {
    extracted.summary = params.summary;
    extracted.lastAssistant = params.summary;
  }
  if (params.toolResults?.length) extracted.toolResults = [...params.toolResults];
  if (params.notes) extracted.notes = params.notes;

  const dodItems = (params.dodItems ?? []).map((item) => item.trim()).filter(Boolean);
  const testCases = (params.testCases ?? []).map((item) => item.trim()).filter(Boolean);
  const verificationMappings = (params.verificationMappings ?? [])
    .map((mapping) => sanitizeVerificationMapping(mapping))
    .filter((mapping): mapping is VerificationMapping => Boolean(mapping));
  if (dodItems.length || testCases.length || verificationMappings.length) {
    extracted.verification = {
      dodItems,
      testCases,
      mappings: verificationMappings,
    };
  }

  const outDir = path.join(paths.evidenceDir, params.taskId);
  fs.mkdirSync(outDir, { recursive: true });
  const outPath = path.join(outDir, `${params.agentId}.json`);
  const payload = {
    ts: nowIso(), taskId: params.taskId, agentId: params.agentId, unitId: params.unitId ?? null,
    runId: params.runId ?? '', sessionKey: params.sessionKey ?? null, sessionId: params.sessionId ?? null,
    transcriptPath: params.transcript ? path.resolve(params.transcript) : null, evidence: extracted,
  };
  fs.writeFileSync(outPath, `${JSON.stringify(payload, null, 2)}\n`, 'utf8');
  appendEvent(paths.eventsFile, {
    event: 'evidence_collected', taskId: params.taskId, phase: 'verification', actor: params.agentId, status: 'IN_PROGRESS', unitId: params.unitId ?? null,
    sessionKey: params.sessionKey ?? null, payload: { agentId: params.agentId, runId: params.runId ?? '', evidencePath: outPath },
  });
  rebuildIndex(paths);
  return { ok: true, taskId: params.taskId, agentId: params.agentId, evidencePath: outPath, sessionId: params.sessionId ?? null, unitId: params.unitId ?? null };
}

export function mergeEvidence(paths: ZigrixPaths, params: { taskId: string; requiredAgents?: string[]; requireQa?: boolean }): Record<string, unknown> | null {
  const task = loadTask(paths, params.taskId);
  if (!task) return null;
  const taskDir = path.join(paths.evidenceDir, params.taskId);
  fs.mkdirSync(taskDir, { recursive: true });
  const files = fs.readdirSync(taskDir).filter((name) => name.endsWith('.json') && name !== '_merged.json').sort();
  const items = files.flatMap((file) => {
    try {
      const data = JSON.parse(fs.readFileSync(path.join(taskDir, file), 'utf8')) as Record<string, unknown>;
      const agentId = String(data.agentId ?? path.basename(file, '.json'));
      return [{ agentId, unitId: data.unitId, runId: data.runId, sessionKey: data.sessionKey, sessionId: data.sessionId, transcriptPath: data.transcriptPath, evidence: data.evidence ?? {} }];
    } catch {
      return [];
    }
  });
  const presentAgents = [...new Set(items.map((item) => String(item.agentId)))].sort();
  const requiredAgents = [...(params.requiredAgents?.length ? params.requiredAgents : resolveRequiredAgents(task))];
  const missingAgents = requiredAgents.filter((agentId) => !presentAgents.includes(agentId));
  const qaAgentId = resolveQaAgentId(task);
  const qaPresent = qaAgentId ? presentAgents.includes(qaAgentId) : false;
  const qaRequiredSatisfied = !(params.requireQa ?? false) || (qaAgentId !== null && qaPresent);

  const qaVerificationMappings = qaAgentId
    ? items
      .filter((item) => String(item.agentId) === qaAgentId)
      .flatMap((item) => collectVerificationMappings(item.evidence as Record<string, unknown>))
    : [];
  const qaVerification = {
    required: Boolean(params.requireQa),
    mappingCount: qaVerificationMappings.length,
    mappings: qaVerificationMappings,
    complete: !(params.requireQa ?? false) || !qaPresent || qaVerificationMappings.length > 0,
  };

  const complete = missingAgents.length === 0 && qaRequiredSatisfied && qaVerification.complete;
  const merged = {
    ts: nowIso(),
    taskId: params.taskId,
    requiredAgents,
    presentAgents,
    missingAgents,
    qaAgentId,
    qaPresent,
    qaVerification,
    complete,
    items,
  };
  const outPath = path.join(taskDir, '_merged.json');
  fs.writeFileSync(outPath, `${JSON.stringify(merged, null, 2)}\n`, 'utf8');
  appendEvent(paths.eventsFile, {
    event: 'evidence_merged', taskId: params.taskId, phase: 'verification', actor: 'zigrix', status: complete ? 'DONE_PENDING_REPORT' : 'IN_PROGRESS',
    payload: {
      requiredAgents,
      missingAgents,
      complete,
      mergedPath: outPath,
      qaPresent,
      qaVerificationComplete: qaVerification.complete,
      qaVerificationMappingCount: qaVerification.mappingCount,
    },
  });
  rebuildIndex(paths);
  return {
    ok: true,
    taskId: params.taskId,
    complete,
    missingAgents,
    qaAgentId,
    qaPresent,
    qaVerificationComplete: qaVerification.complete,
    qaVerificationMappingCount: qaVerification.mappingCount,
    mergedPath: outPath,
  };
}
